### ISSUE TITLE
android: offline downloads for Jellyfin/Emby/Plex songs, albums and playlists (port from Shuttle Podcasts)

### ISSUE THREAD

**Issue by Wigsinator** [timusus/Shuttle2#88]
## 💬  User story

```gerkin
As a user
I want to be able to download tracks from media providers
So that I can listen to them without expending data, or while entirely offline.
```

## 📝 Notes

Ideally, I'd like it to work similarly to spotify's implementation, where a playlist or album can be selected for download, and it will download all tracks within that playlist or album.

**Comment by shabby-goddess** [comment]
Agreed. Downloading is an essential feature for many, especially those who listen on their commutes, since many subways, highways, etc. do not have stable cell connection.

**Comment by timusus** [comment]
Prioritised by the owner (23 Sep 2026) as the next big item after widgets. It's part of strategy Bet 3 (#233), and #83 (Emby sync) overlaps.

Goal: download songs, albums or playlists from a streaming provider and play them offline, with the local file used in place of the stream when it exists.

Reuse from Shuttle Podcasts (`/Users/tim/projects/simplecity-apps/podcasts/main/mobile/android`), which already ships downloads:
- `downloads/` module: `DefaultPodcastDownloadManager`, `DefaultDownloadRepository`, `DownloadRequirementsManager` (Wi-Fi/charging/storage conditions), `PlayedCacheSweeper`
- `data/.../downloads/AutoDownloadCleanupCoordinator`, `PlayedCacheRetention`, `data/.../sync/RunAutoDownloadPass`
- `core/ui/.../DownloadButton`, `PendingDownloadRemovals`

S2-specific work: a per-provider authenticated download URL (Jellyfin/Emby/Plex), Room state for downloaded tracks, a playback path that prefers the local file, UI for download/remove on an album or playlist, and storage management. Not yet scoped against the S2 codebase (unverified how cleanly the podcasts module lifts across).

**Comment by timusus** [comment]
## Port plan (research 2026-09-24, code reading only)

**Reuse from Shuttle Podcasts** (`podcasts/main/mobile/android/downloads/`), port Metro DI → Hilt:
- `di/DownloadsModule.kt` — non-evicting Media3 `SimpleCache` in filesDir (rebuilt on a corrupt index), `DownloadManager` running 2 downloads at once, Wi-Fi-only by default.
- `service/PodcastDownloadService.kt` + `SafeScheduler.kt` — foreground `DownloadService` that handles Android 12+ foreground-start failures; port almost unchanged.
- `DefaultPodcastDownloadManager` (facade, swallows background-start exceptions), `DownloadRequirementsManager`, `DefaultDownloadRepository` + mappers (DownloadIndex as Flows).
- Pattern for sweeping orphaned downloads: `PlayedCacheSweeper`, `AutoDownloadCleanupCoordinator`. The UI pieces `DownloadButton.kt` and `PendingDownloadRemovals.kt` are for Compose album detail only.
- Skip: ad seams, `PlaybackOpenKeys`, auto-download.

**Architecture**
- New `:android:downloads` module on Media3 `DownloadManager`/`DownloadService` (not WorkManager). Separate never-evicted `SimpleCache` in `filesDir/downloads`.
- Cache key = `song.path` (stable across re-imports; the URL changes on every call and carries a token).
- New per-provider `downloadUri(song)` for the original file. Jellyfin/Emby: `/Items/{id}/Download`, falling back to `/Audio/{id}/stream?static=true` (unverified; probe with `support/scripts/media-server-stream-probe.sh`). Plex: its existing path.
- Playback: a lookup interface in `:mediaprovider:core`. The media resolver prefers a completed download (local, cache key) over the stream. The player wraps its data source in a read-only `CacheDataSource`, so streams never write to the cache and avoid podcasts' lock contention. Cast keeps streaming.
- DB: v1 needs no change to the `songs` table, since state lives in Media3's DownloadIndex. Pinning an album or playlist would add a `pinned_collections` table (Room 41→42).
- UI in existing screens only: Compose AlbumDetail toolbar, MVP playlist detail menu, rows in `preferences_playback.xml`.
- #55: a "Network and data" group with a Wi-Fi-only download toggle and a cellular max-bitrate setting for Jellyfin/Emby `/universal`.

**Slices** (each lands on its own; 1 and 3 wait for the Media3 migration #304)
1. Download infrastructure module, service, Hilt, and a debug receiver (hard)
2. Per-provider download URLs + tests (standard)
3. Playback prefers the download; check playback in airplane mode on the emulator (hard)
4. UI entry points in album detail and the playlist menu, with robot tests (standard)
5. Settings: Wi-Fi-only, space used, remove all, orphan sweeper (standard)
6. #55 cellular bitrate cap (standard)

**Risks:** Android 15+ caps `dataSync` foreground services at 6h/day; queued requests store token URLs, which will 401 after the token rotates; Jellyfin `/Download` needs the user's download permission; offline, songs that aren't downloaded are still listed and fail to play; recovering from a corrupt index wipes every download.

**Open questions for owner**
1. Original quality only, or transcoded downloads too?
2. Pinned albums/playlists that auto-download new tracks (needs the migration), or one-off downloads for v1?
3. Which Jellyfin/Emby endpoint (probe the test servers)?
4. A "Downloaded only" filter or offline mode, and when?
5. Is filesDir OK, or a user-chosen folder?

**Comment by timusus** [comment]
Owner decisions 2026-09-24:
- **Pinning:** yes. Downloading an album or playlist pins it, and new tracks auto-download after each import, the way Spotify does it. This needs the `pinned_collections` Room migration (41→42) in slice 1 or 5.
- **Quality:** original file only.
- **Storage:** follow Spotify: app-private storage by default (`getExternalFilesDirs`/filesDir, deleted on uninstall, no permissions needed), with an internal vs SD-card choice later if a removable volume is present.

**Comment by timusus** [comment]
Slice 1 landed (c5322c3d, 5265b914, c5419ca7, f4752298): the :android:downloads engine, the pinned_collections table (Room 42→43), a debug download receiver and main-thread DownloadManager construction. Not user-visible yet. Note for the UI slice: the app never requests POST_NOTIFICATIONS at runtime (the only runtime permission request is StoragePermissionScreen.kt), so download notifications won't show on Android 13+ until the UI slice asks for it. Slice 2 (download URLs) is in progress.
